### PR TITLE
Miscellaneous bug fixes and enhancements for the new HDF5 IO type

### PR DESCRIPTION
**Pull request by dqwu**
This PR updated the nightly unit test for the new HDF5 IO type and
fixed some issues found in E3SM production runs.

Planned code refactoring work will be performed later, including
more error handling and helper functions.

### CODE REVIEW

File: examples/c/test_hdf5.c [hunk 1]
```diff
@@ -72,216 +91,485 @@ int main(int argc, char* argv[])
...
+    }
 
     MPI_Finalize();
 
 #ifdef TIMING
     GPTLfinalize();
 #endif
 
+#else
+    printf("SCORPIO is not configured with HDF5 support.\n");
```
Comment (jayeshkrishna) [inline examples/c/test_hdf5.c:571]: In a future PR, make the test fail if HDF5 support is not enabled/available. In cmake also ensure that this test is only included/run if direct HDF5 support is enabled.

File: src/clib/pio_getput_int.c [hunk 2]
```diff
@@ -307,6 +307,30 @@ int PIOc_put_att_tc(int ncid, int varid, const char *name, nc_type atttype,
 #ifdef _HDF5
     if (file->iotype == PIO_IOTYPE_HDF5)
     {
+        int num_attrs = file->hdf5_num_attrs;
+        if (num_attrs >= PIO_MAX_ATTRS)
+        {
+            fprintf(stderr, "ERROR: Num of attributes exceeds maximum (%d).\n", PIO_MAX_ATTRS);
```
Comment (jayeshkrishna) [inline src/clib/pio_getput_int.c:313]: We might want to refrain from direct fprintfs and let pio_err handle the error message

File: src/clib/pio_nc.c [hunk 3]
```diff
@@ -1869,6 +1958,37 @@ int PIOc_inq_att(int ncid, int varid, const char *name, nc_type *xtypep,
     }
 #endif
 
+#ifdef _HDF5
+    if (file->iotype == PIO_IOTYPE_HDF5)
+    {
+        ierr = PIO_ENOTATT;
+        for (int i = 0; i < file->hdf5_num_attrs; i++)
+        {
+            if (!strcmp(name, file->hdf5_attrs[i].att_name) &&
```
Comment (jayeshkrishna) [inline src/clib/pio_nc.c:1967]: In future PR don't forget to replace strcmps by strncmps

**Review by jayeshkrishna — APPROVED** [review]

File: src/clib/pio_nc.c (hunk 3 above)
Comment (dqwu) [reply src/clib/pio_nc.c:1967]: OK, we can also use strncmps for ADIOS type.

File: examples/c/test_hdf5.c (hunk 1 above)
Comment (dqwu) [reply examples/c/test_hdf5.c:571]: OK, thanks for the suggestion.

File: src/clib/pio_getput_int.c (hunk 2 above)
Comment (dqwu) [reply src/clib/pio_getput_int.c:313]: OK, ADIOS type has the same issue and we will fix it for both types.

**Pull request merged**